src/back/hlsl/writer.rs

```diff
@@ -712,7 +712,20 @@ impl<'a, W: fmt::Write> super::Writer<'a, W> {
         let struct_name = &self.names[&NameKey::Type(handle)];
         writeln!(self.out, "struct {} {{", struct_name)?;
 
+        let mut last_offset = 0;
         for (index, member) in members.iter().enumerate() {
+            if member.binding.is_none() && member.offset > last_offset {
+                // using int as padding should work as long as the backend
+                // doesn't support a type that's less than 4 bytes in size
+                // (Error::UnsupportedScalar catches this)
+                let padding = (member.offset - last_offset) / 4;
+                for i in 0..padding {
+                    writeln!(self.out, "{}int _pad{}_{};", back::INDENT, index, i)?;
```

Comment:
If we add this padding, all of the accesses into the structure would be broken. I.e. we need to handle indexing, know where padding was inserted, and modify the indices accordingly.

Reply:
The [constructor function](https://github.com/gfx-rs/naga/blob/7d0644589755efefd7c6c7307ba59418f7af6f84/src/back/hlsl/help.rs#L355) doesn't require explicit initialization for all members.

I'm not sure what indexing means in this context since it's a struct.

Reply:
Ah, I think I know what you mean by adjusting the indices. Structs also use `Expression::AccessIndex` which uses an index for accessing struct members.

This is not an issue though since we are just adding the padding without any other part of the code knowing about it. As mentioned previously we can get away with it since we don't need to initialize the padding explicitly.

Reply:
Sounds about right. This is more complicated in MSL, but fortunately not here.

Reply:
Indeed, I first used the same approach as in the MSL backend only to discover that it wasn't necessary.